### ISSUE TITLE
[QUERY] Unable to SqlClient instrumentation

### ISSUE THREAD

**Issue by rouen-sk** [Azure/azure-sdk-for-net#41530]
### Library name and version

Azure.Monitor.OpenTelemetry.AspNetCore 1.0.0

### Query/Question

I just upgraded Azure.Monitor.OpenTelemetry.AspNetCore from beta5 to 1.0.0. and noticed the breaking change - classes like `SqlClientInstrumentationOptions` and `AspNetCoreInstrumentationOptions` are now internal instead of public.

While on beta, I was using these classes to setup OpenTelemetry in my ASP.NET (.NET 8) project, like this:

```
public static IServiceCollection AddOpenTelemetryCustomized(this IServiceCollection services, ConfigurationManager configuration)
{
	services.AddOptions<SqlClientInstrumentationOptions>().Configure(options =>
	{
		configuration.Bind("OpenTelemetry:Sql", options); // configuration for SQL statements included in traces
		options.Filter = TelemetryFilters.ShouldCollectSql; // exclude health check dummy queries
	});

	services.AddOptions<AspNetCoreInstrumentationOptions>().Configure(options =>
	{
		// exclude things like healthchecks, swagger, visual studio debug endpoints
		options.Filter = TelemetryFilters.ShouldCollectRequest;
		options.EnrichWithHttpRequest = EnrichWithHttpRequestAction;
		options.EnrichWithHttpResponse = EnrichWithHttpResponseAction;
	});

	// ApplicationInsights connection string is expected in enviroment variable
	services.AddOpenTelemetry().UseAzureMonitor().ConfigureResource(builder =>
	{
		builder.AddService(serviceName);
	});
}
```
Now that I can't use them, what is the clean/recommended way to configure these things (for example getting value for `SetDbStatementForText` from environment variable, if present, etc.). 
Thanks!

### Environment

.NET 8

**Comment by cijothomas** [comment]
Asp.Net Core, HttpClients were vendored in, but not anymore, so you should be able to get it working.
For SQL, you need to add explicit dependency to https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Instrumentation.SqlClient and then try again, and it should work.

https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore#customizing-sqlclientinstrumentationoptions

**Comment by cijothomas** [comment]
@rajkumar-rangaraj the public docs is absent of this information! Need to add more content from this repo to docs for easy discovery.

**Comment by rouen-sk** [comment]
Thank you @cijothomas but I am still bit confused.
I have to add package references to both `OpenTelemetry.Instrumentation.AspNetCore` (currently 1.7.0) and `OpenTelemetry.Instrumentation.SqlClient` (currently 1.6.0-beta.3) to have access to classes `SqlClientInstrumentationOptions` and `AspNetCoreTraceInstrumentationOptions` (renamed from `AspNetCoreInstrumentationOptions` I guess). 

But if I understand [this docs](https://github.com/Azure/azure-sdk-for-net/blob/Azure.Monitor.OpenTelemetry.AspNetCore_1.0.0/sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md#troubleshooting) and code in method `AddVendorInstrumentationIfPackageNotReferenced` correctly, that will effectively prevent `UseAzureMonitor()` from adding both AspNetCore and SqlClient instrumentation, and I have to call them explicitly like this:

```
services.AddOpenTelemetry().UseAzureMonitor()
    .WithTracing(builder =>
    {
        builder.AddAspNetCoreInstrumentation(options =>
        {
            // exclude things like healthchecks, swagger, visual studio debug endpoints
            options.Filter = TelemetryFilters.ShouldCollectRequest;
            options.EnrichWithHttpRequest = EnrichWithHttpRequestAction;
            options.EnrichWithHttpResponse = EnrichWithHttpResponseAction;
        });

        builder.AddSqlClientInstrumentation(options =>
        {
            configuration.Bind("OpenTelemetry:Sql", options); // configuration for SQL statements included in traces
            options.Filter = TelemetryFilters.ShouldCollectSql; // exclude health check dummy queries
        });
    })
    .ConfigureResource(builder =>
    {
        builder.AddService(serviceName);
    });
```
So basically, the behavior is: if you want all options on defaults, `UseAzureMonitor()` will add instrumentations, but if you want to customize some options, you need to add instrumentations manually (instead of them being called from `UseAzureMonitor` and "picking up" on configured options like it was before). Is that correct? 

Note: I wanted to link also source code of `AddVendorInstrumentationIfPackageNotReferenced`, but code here on GitHub does not match what I see when I F12 `UseAzureMonitor` (from `Azure.Monitor.OpenTelemetry.AspNetCore` 1.0.0. package in my project). Dictionary look like this, so it disables both AspNetCore and SqlClient instrumentation:
```
var vendorInstrumentationActions = new Dictionary<string, Action>
{
    { AspNetCoreInstrumentationPackageName, () => tracerProviderBuilder.AddAspNetCoreInstrumentation() },
    { SqlClientInstrumentationPackageName, () => tracerProviderBuilder.AddSqlClientInstrumentation() },
    {
        HttpClientInstrumentationPackageName,
            () => tracerProviderBuilder.AddHttpClientInstrumentation(o => o.FilterHttpRequestMessage = (_) =>
                {
                // Azure SDKs create their own client span before calling the service using HttpClient
                // In this case, we would see two spans corresponding to the same operation
                // 1) created by Azure SDK 2) created by HttpClient
                // To prevent this duplication we are filtering the span from HttpClient
                // as span from Azure SDK contains all relevant information needed.
                var parentActivity = Activity.Current?.Parent;
                if (parentActivity != null && parentActivity.Source.Name.Equals("Azure.Core.Http"))
                {
                    return false;
                }

                return true;
                })
    },
};
```

**Comment by TimothyMothra** [comment]
Hi @rouen-sk,
You caught us in a middle of a release cycle, so the code in the `main` branch of this repo doesn't exactly match some of the docs. I'm deeply sorry for that.

To find the source code for a specific release, you can browse the Releases in this repo and click on the commit id. https://github.com/Azure/azure-sdk-for-net/releases?q=AspNetCore&expanded=true

--
Regarding our versions and dependencies, let me try to explain what's happening..

- While `OpenTelemetry.Instrumentation.AspNetCore` was in beta, we had public dependencies of all three Instrumentation libraries which were also in beta. (_AspNetCore, Http, and SqlClient_).

- When `OpenTelemetry.Instrumentation.AspNetCore` shipped v1.0.0, we couldn't ship with beta dependencies, so the source code for all 3 Instrumentation libraries (_AspNetCore, Http, and SqlClient_) was privately vendored in our sdk. The `UseAzureMonitor()` applies only the default options of the Instrumentation libraries. If you need to customize either of these Instrumentation libraries you would need to manually onboard to the Instrumentation libraries to access their public option APIs. In this scenario, our library would backoff to allow your configuration to take precedence. (This statement is missing from our public docs).

- I'm currently readying `OpenTelemetry.Instrumentation.AspNetCore` v1.1.0 which will be released in days. This will include package dependencies to the public stable _AspNetCore_ and _Http_ Instrumentation libraries and you'll be able to use their public options APIs without any extra effort. This won't help you for the _SqlClient_ instrumentation which is still in beta and will still require manual onboarding to configure its options. (This statement is available now in the repo's [Readme](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md#customizing-instrumentation-libraries))


Please let me know if you're still confused and I can explain further.

**Comment by rouen-sk** [comment]
Hi @TimothyMothra thank you for explanation. I am OK with the code above (calling `UseAzureMonitor()` and then also `AddAspNetCoreInstrumentation` and `AddSqlClientInstrumentation`, no big deal, I was just curious, if I am supposed to do so or I am missing something. 

I still think it is unusual design decision and kinda unexpected behavior - ASP.NET itself and many 3rd party libraries which support registration to DI container via extension methods simply just have their Option class public, and you can `Configure` it into container and that does not mean you suddenly have to call extra service registration methods, because default one is "backing off" just because you have configured options. But I guess there must be reason for it I don't see :)
